Fix the SaferCPP bot build after 286031@main

geoffreygaren opened the pull request:
#### 238d5dac981a79795488c0b33d6d83b096629e7a
<pre>
Fix the SaferCPP bot build after 286031@main
<a href="https://bugs.webkit.org/show_bug.cgi?id=282483">https://bugs.webkit.org/show_bug.cgi?id=282483</a>
&lt;<a href="https://rdar.apple.com/problem/139113248">rdar://problem/139113248</a>&gt;

Reviewed by Chris Dumez.

Resolved some unsafe spans in the !HAVE(SWIFT_CPP_INTEROP) code path.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::deriveBitsCoreCrypto):
* Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
(WebCore::CryptoKeyEC::platformExportSpki const):
(WebCore::CryptoKeyEC::platformExportPkcs8 const):
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::generateJwkX const):

Canonical link: <a href="https://commits.webkit.org/286304@main">https://commits.webkit.org/286304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ff4febb63b0cff4a42878fa8961633bddb1c2d26

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75558 "Passed style check") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/55/builds/28409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/26822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77674 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/2789 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/80038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/26822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78625 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/55/builds/28409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/80038 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/55/builds/28409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25150 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/220 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/28409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81517 "Built successfully") | 
| | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/2897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/1828 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/67508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/3048 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/55/builds/28409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/66802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/55/builds/28409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11664 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/2854 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/5676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/2879 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/3814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/2886 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nmahendru reviewed (approved):
lgtm! thanks!

cdumez commented on Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
```diff
@@ -364,6 +364,7 @@ Vector<uint8_t> CryptoKeyEC::platformExportSpki() const
     // <rdar://problem/30987628>
     const uint8_t* oid;
     size_t oidSize = getOID(namedCurve(), oid);
+    auto oidSpan = unsafeMakeSpan(oid, oidSize);
```
Ideally we'd fix getOID() instead to return a span.

cdumez commented on Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
```diff
@@ -380,7 +380,7 @@ RefPtr<CryptoKeyOKP> CryptoKeyOKP::importPkcs8(CryptoAlgorithmIdentifier identif
         return nullptr;
 
-    return create(identifier, namedCurve, CryptoKeyType::Private, std::span { keyData.data() + index, keyData.size() - index }, extractable, usages);
+    return create(identifier, namedCurve, CryptoKeyType::Private, keyData.span().subspan(index, keyData.size() - index), extractable, usages);
```
`keyData.subSpan()` should work iirc.

cdumez commented on Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
```diff
@@ -312,16 +304,15 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
     if (keyData.size() < index + sizeof(IdEcPublicKey))
         return nullptr;
-    if (memcmp(keyData.data() + index, IdEcPublicKey, sizeof(IdEcPublicKey)))
+    if (!equalSpans(keyData.subspan(index), std::span { IdEcPublicKey }))
```
This looks risky, are we sure that `keyData.subspan(index)` has the same size as `IdEcPublicKey`? The previous code would behavior differently if the sizes were different, since it didn't check the size.

`keyData.subspan(index, IdEcPublicKey.size())` would be safer for e.g., unless you're confident it'd fine as-is

geoffreygaren replied:
Oh, good point! Somehow I implicitly assumed that equalSpans would pick the smaller size, but that's obviously not the case.

cdumez commented on Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
```diff
@@ -312,16 +304,15 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportSpki(CryptoAlgorithmIdentifier id
+    if (keyData.size() < index + oid.size())
         return nullptr;
-    if (memcmp(keyData.data() + index, oid, oidSize))
+    if (!equalSpans(keyData.subspan(index), oid))
```
ditto.

cdumez commented on Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
```diff
@@ -403,16 +393,15 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportPkcs8(CryptoAlgorithmIdentifier i
     if (keyData.size() < index + sizeof(IdEcPublicKey))
         return nullptr;
-    if (memcmp(keyData.data() + index, IdEcPublicKey, sizeof(IdEcPublicKey)))
+    if (!equalSpans(keyData.subspan(index), std::span { IdEcPublicKey }))
```
ditto.

cdumez commented on Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
```diff
@@ -403,16 +393,15 @@ RefPtr<CryptoKeyEC> CryptoKeyEC::platformImportPkcs8(CryptoAlgorithmIdentifier i
+    if (keyData.size() < index + oid.size())
         return nullptr;
-    if (memcmp(keyData.data() + index, oid, oidSize))
+    if (!equalSpans(keyData.subspan(index), oid))
```
ditto.

geoffreygaren committed:
Fix the SaferCPP bot build after 286031@main
https://bugs.webkit.org/show_bug.cgi?id=282483
<rdar://problem/139113248>

Reviewed by Chris Dumez.

Resolved some unsafe spans in the !HAVE(SWIFT_CPP_INTEROP) code path.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/crypto/cocoa/CryptoAlgorithmEd25519Cocoa.cpp:
(WebCore::signEd25519):
* Source/WebCore/crypto/cocoa/CryptoAlgorithmX25519Cocoa.cpp:
(WebCore::deriveBitsCoreCrypto):
* Source/WebCore/crypto/cocoa/CryptoKeyECMac.cpp:
(WebCore::CryptoKeyEC::platformExportSpki const):
(WebCore::CryptoKeyEC::platformExportPkcs8 const):
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::importSpki):
(WebCore::CryptoKeyOKP::importPkcs8):
(WebCore::CryptoKeyOKP::generateJwkX const):

Canonical link: https://commits.webkit.org/286304@main
